### PR TITLE
Add run time info and upp

### LINKED ISSUE #677: Adding run time information for fv3 grid component
## Description
This PR is to add run time information at each time step in fv3 cap for the NUOPC phases and upp updates.

## Solution
1. Add timing information in the fv3 cap.
2. For UPP: 
(1) make upp revision up-to-date
(2) add printout information for inline post 2d decomposition
(3) tweak reading microphysics number concentration
(4) read 3D radar reflectivity when model uses Thmopson MP

## Alternatives
If applicable, add a description of any alternative solutions or features you've considered.

## Testing:

1. Have you tested the code changes? On what platforms?
on hera
2. Have you run regression test in ufs-weather-model code changes?
- Will the baseline results change?
- If yes, please give brief explanation of your code updates on the regression test.

## Dependent PRs:

Directly reference any issues or PRs in this or other repositories that this is related to, and describe how they are related. Example:
- associated https://github.com/ufs-community/ufs-weather-model/pull/1823

### LINKED ISSUE #1031: unify the filled_value in atmosphere history fiels
## Description
Current UFS output uses different missing values (FillValue) in atmosphere 3D field files and the surface files. In the project of unifying the post interfaces for all the applications, we need to first update those files to use a unified missing value in all the history files for all the applications.   

## Solution
Use a unified missing value so that POST processing can use the single missing value to identify grid points with missing value.

## Related to
Directly reference any issues or PRs in this or other repositories that this is related to, and describe how they are related. Example:
- required to support noaa-emc/nems/issues/<issue_number>
- needed by noaa-emc/fv3atm/pull/<pr_number>

### PR DESCRIPTION
**Pull request by junwang-noaa**
## Description

This PR is to add run time info and using consistent missing value in fv3atm.

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues should always be created before starting work on a PR branch!)
- fixes https://github.com/NOAA-EMC/fv3atm/issues/677
- fixes https://github.com/ufs-community/ufs-weather-model/issues/1031

## Testing

How were these changes tested?  
What compilers / HPCs was it tested with?  
Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)  
Have the ufs-weather-model regression test been run? On what platform?  
- Will the code updates change regression test baseline? If yes, why? Please show the baseline directory below.
- Please commit the regression test log files in your ufs-weather-model branch

## Dependencies

If testing this branch requires non-default branches in other repositories, list them.
Those branches should have matching names (ideally)

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs
- waiting on noaa-emc/nems/pull/<pr_number>
- waiting on noaa-emc/fv3atm/pull/<pr_number>

### CODE REVIEW

File: fv3_cap.F90 [hunk 1]
```diff
@@ -211,8 +215,6 @@ subroutine InitializeAdvertise(gcomp, rc)
     logical                                :: use_saved_routehandles, rh_file_exist
     logical                                :: fieldbundle_is_restart = .false.
 
-    integer                                :: sloc
-    type(ESMF_StaggerLoc)                  :: staggerloc
```
Comment (DusanJovic-NOAA) [inline fv3_cap.F90]: Why are these deleted?

File: fv3_cap.F90 [hunk 2]
```diff
@@ -636,12 +640,7 @@ subroutine InitializeAdvertise(gcomp, rc)
                 dstGrid(j,i) = grid
                 ! loop over all the mirror fields and set the balanced mirror Grid
                 do ii=1, fieldCount
-                  call ESMF_InfoGetFromHost(fieldList(ii), info=info, rc=rc)
-                  if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
-                  call ESMF_InfoGet(info, key="staggerloc", value=sloc, rc=rc)
-                  if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
-                  staggerloc = sloc  ! convert integer into StaggerLoc_Flag
-                  call ESMF_FieldEmptySet(fieldList(ii), grid=grid, staggerloc=staggerloc, rc=rc)
```
Comment (DusanJovic-NOAA) [inline fv3_cap.F90]: Why are these deleted?

File: fv3_cap.F90 [hunk 3]
```diff
@@ -1011,10 +1017,13 @@ subroutine ModelAdvance(gcomp, rc)
 
     type(ESMF_GridComp)         :: gcomp
     integer, intent(out)        :: rc
+    real(kind=8)                :: MPI_Wtime
 
 !-----------------------------------------------------------------------------
 
     rc = ESMF_SUCCESS
+    timers = MPI_Wtime()
+    if(lprint) print *,'in FV3 cap, time between fv3 run step=', timers-timere,mype
```
Comment (DusanJovic-NOAA) [inline fv3_cap.F90]: Consider changing 'FV3 cap' to 'fv3_cap' to be consistent with other prints

File: fv3_cap.F90 [hunk 4]
```diff
@@ -1026,6 +1035,9 @@ subroutine ModelAdvance(gcomp, rc)
 
     if (profile_memory) call ESMF_VMLogMemInfo("Leaving FV3 ModelAdvance: ")
 
+    timere = MPI_Wtime()
+    if(lprint) print *,'in FV3 cap, time in fv3 run step=', timere-timers, mype
```
Comment (DusanJovic-NOAA) [inline fv3_cap.F90]: Consider changing 'FV3 cap' to 'fv3_cap' to be consistent with other prints

**Comment by GeorgeVandenberghe-NOAA** [comment]
Do you have a sample printout?  Would like to see this at restart write time.

Not a crippling error but it would be helpful to accumulate the run duration sum of timers - timere to look for spikes.  This is how I found restart and history write spikes from various components.

File: fv3_cap.F90 [hunk 5]
```diff
@@ -1040,10 +1052,13 @@ subroutine ModelAdvance_phase1(gcomp, rc)
     logical                     :: fcstpe
     character(len=*),parameter  :: subname='(fv3_cap:ModelAdvance_phase1)'
     character(240)              :: msgString
+    real(kind=8)                :: MPI_Wtime
 
 !-----------------------------------------------------------------------------
 
     rc = ESMF_SUCCESS
+    timep1rs = MPI_Wtime()
+    if(lprint) print *,'in fv3_cap,modeladvance phase1 time ', timep1rs-timep2re,mype
```
Comment (DusanJovic-NOAA) [inline fv3_cap.F90]: Shouldn't this message be something like ' .. time between phase 2 and phase 1' like in ModelAdvance? We are subtracting here start of this subroutine from the end of the previous time step phase 2.

**Comment by GeorgeVandenberghe-NOAA** [comment]
I should note the purpose of this PR is to provide better runtime timestep
statistics, an excellent idea@

On Wednesday, July 5, 2023, Dusan Jovic ***@***.***> wrote:
> @DusanJovic-NOAA commented on this pull request.
>
> ________________________________
>
> In fv3_cap.F90:
>
>>
>
!-----------------------------------------------------------------------------
>
> rc = ESMF_SUCCESS
> + timep1rs = MPI_Wtime()
> + if(lprint) print *,'in fv3_cap,modeladvance phase1 time ',
timep1rs-timep2re,mype
>
> Shouldn't this message be something like ' .. time between phase 2 and
phase 1' like in ModelAdvance? We are subtracting here start of this
subroutine from the end of the previous time step phase 2.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because your review was requested.<
https://ci3.googleusercontent.com/proxy/LlJvHy_vgPRkUyK2Ua262198CtehAztT-WAn4ULUsyv9rCZKnGV6oW29xeiVmlZeyX989v274eJyPJ71dmEVIvfa8ZRfQDrWCjnxW0MJ8ZZ0iL2EzIEsO8YmMU5dyynZx13Ve29fpXbkJrQqWNdlW42lL_qczH4yWnrNZttKK-sfDqpM8TgUVxS7H7jdW2JP64G7e-WQqjCXYrfvQkgRTUGCr20SMP_OZlMiVV5za3r0yuvuHmTvapEKIKIX=s0-d-e1-ft#https://github.com/notifications/beacon/ANDS4FU57Z2RUCW6OM7Z4ITXOXFMVA5CNFSM6AAAAAAZ7NKJGKWGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTS2KHSLS.gif>Message
ID: ***@***.***>

-- 

George W Vandenberghe

*Lynker Technologies at * NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***


301-683-3769(work)  3017751547(cell)

File: fv3_cap.F90 (hunk 2 above)
Comment (junwang-noaa) [reply fv3_cap.F90]: Sorry, I missed the change. Corrected.

File: fv3_cap.F90 (hunk 5 above)
Comment (junwang-noaa) [reply fv3_cap.F90]: Yes. changed.

File: fv3_cap.F90 [hunk 6]
```diff
@@ -1003,6 +1011,11 @@ subroutine InitializeRealize(gcomp, rc)
 
     if (ESMF_LogFoundError(rcToCheck=urc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__, rcToReturn=rc)) return
 
+    timers = 0.
+    timere = 0.
```
Comment (DusanJovic-NOAA) [inline fv3_cap.F90]: timep2re should also be initialized, because on the very first call of  ModelAdvance_phase1 it will be uninitialized. 

Also if these values are initially zero, on the very first call of ModelAdvance* the elapsed time will not be correct. Since MPI_WTime returns time since arbitrary point in the past.  Maybe if these values are zero you can skip printing initial time difference between the current and the previous time.

Comment (junwang-noaa) [reply fv3_cap.F90]: Updated.

Comment (DusanJovic-NOAA) [reply fv3_cap.F90]: Thanks.

Well, strictly speaking all 'start' times (timers, timep1rs, timep2rs) can actually be declared as local variables, they do not have to be saved at the module level, since they are only used locally in three respective 'advance' routines. 

timep1re can also be local in ModelAdvance_phase1. Only timere and timep2re must be saved at the module level.

I think the code is now correct, so these are just suggestions.

Comment (junwang-noaa) [reply fv3_cap.F90]: I was thinking of that too. The code is updated.

**Comment by junwang-noaa** [comment]
> Do you have a sample printout? Would like to see this at restart write time.
> 
> Not a crippling error but it would be helpful to accumulate the run duration sum of timers - timere to look for spikes. This is how I found restart and history write spikes from various components.

@GeorgeVandenberghe-NOAA You can see a test case at: /work/noaa/stmp/junwang/stmp/junwang/FV3_RT/rt_250962/control_p8_intel
It is a low resolution C96 case, the restart file is small compared to radiation time.

**Comment by junwang-noaa** [comment]
@DusanJovic-NOAA would you please review the code again?

**Review by DusanJovic-NOAA — APPROVED** [review]

**Review by WenMeng-NOAA — APPROVED** [review]
The updates of inline post interface and the upp revision look good to me. Thanks!

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Comment by FernandoAndrade-NOAA** [comment]
Ok, all conversations are resolved and testing for #[1823 ](https://github.com/ufs-community/ufs-weather-model/pull/1823)has completed successfully, please continue with the merge process.
